src/Domain/CoordinateMaps/CylindricalEndcap.cpp

```diff
@@ -0,0 +1,914 @@
+  // dz/dxbar
+  get<2, 0>(jacobian_matrix) = d_sin_factor * ybar * xbar;
+  // dz/dybar
+  get<2, 1>(jacobian_matrix) = d_sin_factor * square(ybar) + sin_factor;
+
+  return jacobian_matrix;
+}
+
+// This is really a 2-dimensional inverse jacobian because zbar = -1
+// and xbar,ybar depend only on y,z (given y,z we know x since
+// x,y,z must be on the sphere).
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>
+InnerSphereMap::inv_jacobian(const std::array<T, 3>& source_coords) const
+    noexcept {
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  const return_type& xbar = source_coords[0];
+  const return_type& ybar = source_coords[1];
+  const return_type rho = sqrt(square(xbar) + square(ybar));
+  // Let q = sin(rho theta)/rho
+  const return_type q = sin_ax_over_x(rho, theta_);
+  const return_type dlogrho_q = dlogx_sin_ax_over_x(rho, theta_);
+  const return_type one_over_r_q = 1.0 / (q * radius_);
+  const return_type tmp =
+      one_over_r_q * dlogrho_q / (q + square(rho) * dlogrho_q);
+
+  auto inv_jacobian_matrix =
+      make_with_value<tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>>(
+          dereference_wrapper(source_coords[0]), 0.0);
+
+  // dxbar/dy
+  get<0, 1>(inv_jacobian_matrix) = one_over_r_q - square(xbar) * tmp;
+  // dxbar/dz
+  get<0, 2>(inv_jacobian_matrix) = -xbar * ybar * tmp;
+  // dybar/dy
+  get<1, 1>(inv_jacobian_matrix) = get<0, 2>(inv_jacobian_matrix);
+  // dybar/dz
+  get<1, 2>(inv_jacobian_matrix) = one_over_r_q - square(ybar) * tmp;
+
+  return inv_jacobian_matrix;
+}
+
+// Should this go into QuadraticEquation.hpp?
+double positive_root_greater_than_or_equal_to_one(double a, double b,
+                                                  double c) noexcept {
+  const auto roots = real_roots(a, b, c);
+  // Roots are returned in increasing order.
+  if (roots[0] >= 1.0 or equal_within_roundoff(roots[0], 1.0)) {
+    return roots[0];
+  }
+  ASSERT(roots[1] >= 1.0 or equal_within_roundoff(roots[1], 1.0),
+         "No root >=1.  Roots are " << roots[0] << " and " << roots[1]
+                                    << ", with a=" << a << " b=" << b
+                                    << " c=" << c);
+  return roots[1];
+}
+
+// Should this go into QuadraticEquation.hpp?
+DataVector positive_root_greater_than_or_equal_to_one(DataVector a,
+                                                      DataVector b,
+                                                      DataVector c) noexcept {
+  ASSERT(a.size() == b.size(),
+         "Size mismatch a vs b: " << a.size() << " " << b.size());
+  ASSERT(a.size() == c.size(),
+         "Size mismatch a vs c: " << a.size() << " " << c.size());
+  DataVector result(a.size());
+  for (size_t i = 0; i < a.size(); ++i) {
+    result[i] = positive_root_greater_than_or_equal_to_one(a[i], b[i], c[i]);
+  }
+  return result;
+}
+
+// Consider a sphere with center 'sphere_center' and radius 'radius',
```

Comment:
This is a general way of constructing maps, maybe it can be moved out into file containing helper functions for maps.

Reply:
Yes, I was thinking of doing this for the next PR when generalizing.